File: MyAFBase/MyAFBase/Services/AFISearch/AFISearchIndex.swift
import Foundation
import SQLite3

enum AFISearchIndexError: Error {
    case openFailed
    case prepareFailed
    case executionFailed
    case missingEmbedding
}

actor AFISearchIndex {
    private let databaseURL: URL
    private var database: OpaquePointer?
    private var indexedCorpusVersion: String?

    init(databaseURL: URL? = nil) {
        if let databaseURL {
            self.databaseURL = databaseURL
        } else {
            let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            self.databaseURL = appSupport
                .appendingPathComponent("MyAFBase/AFISearch", isDirectory: true)
                .appendingPathComponent("search_index.sqlite", isDirectory: false)
        }
    }

    deinit {
        if let database {
            sqlite3_close(database)
        }
    }

    /// Builds the keyword (FTS) index. This is fast — a few seconds even for the full
    /// corpus — and search is usable as soon as it completes. Semantic embeddings are
    /// filled in afterwards via `backfillEmbeddings`.
    func rebuildIfNeeded(corpus: AFICorpus, progress: (@Sendable (Double) -> Void)? = nil) throws {
        try openIfNeeded()

        if indexedCorpusVersion == corpus.version,
           try chunkCount() == corpus.chunks.count {
            return
        }

        try rebuild(corpus: corpus, progress: progress)
        indexedCorpusVersion = corpus.version
    }

    /// Number of chunks that still need a semantic embedding.
    func pendingEmbeddingCount() throws -> Int {
        try openIfNeeded()

        let statement = try prepare("SELECT COUNT(*) FROM chunks WHERE length(embedding) = 0;")
        defer { sqlite3_finalize(statement) }

        guard sqlite3_step(statement) == SQLITE_ROW else { return 0 }
        return Int(sqlite3_column_int(statement, 0))
    }

    /// Computes semantic embeddings for chunks that don't have one yet, in parallel
    /// batches. Each batch commits separately, so an interrupted run (app killed,
    /// background time expired) resumes from where it stopped instead of restarting.
    func backfillEmbeddings(
        batchSize: Int = 256,
        progress: (@Sendable (Double) -> Void)? = nil
    ) async throws {
        try openIfNeeded()

        let totalPending = try pendingEmbeddingCount()
        guard totalPending > 0 else {
            progress?(1)
            return
        }

        var completed = 0

        while true {
            try Task.checkCancellation()

            let batch = try pendingEmbeddingBatch(limit: batchSize)
            guard !batch.isEmpty else { break }

            let texts = batch.map { AFISearchIndex.ftsDocumentText(for: $0) }
            let vectors = await AFIEmbeddingService.embedAll(texts)

            try execute("BEGIN IMMEDIATE TRANSACTION;")
            do {
                let update = try prepare("UPDATE chunks SET embedding = ? WHERE id = ?;")
                defer { sqlite3_finalize(update) }

                for (index, chunk) in batch.enumerated() {
                    // A 1-byte sentinel marks "attempted but unavailable" so failed
                    // chunks aren't retried forever; search safely ignores it.
                    let data = vectors[index].isEmpty
                        ? AFISearchIndex.embeddingUnavailableSentinel
                        : vectors[index]
                    try bindBlob(update, index: 1, data)
                    try bindText(update, index: 2, chunk.id)
                    try stepDone(update)
                    sqlite3_reset(update)
                }

                try execute("COMMIT;")
            } catch {
                try? execute("ROLLBACK;")
                throw error
            }

            completed += batch.count
            progress?(min(1, Double(completed) / Double(totalPending)))
        }
    }

    func search(
        query: String,
        limit: Int = 12
    ) throws -> [(chunk: AFICorpusChunk, keywordScore: Double, semanticScore: Double)] {
        try openIfNeeded()

        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return [] }

        let keywordHits = try keywordSearch(query: trimmed, limit: max(limit * 4, 24))
        let semanticHits = try semanticSearch(query: trimmed, limit: max(limit * 4, 24))

        var merged: [String: (chunk: AFICorpusChunk, keywordScore: Double, semanticScore: Double)] = [:]

        for (rank, hit) in keywordHits.enumerated() {
            let rrf = 1.0 / Double(60 + rank + 1)
            var entry = merged[hit.chunk.id] ?? (hit.chunk, 0, 0)
            entry.keywordScore += rrf
            merged[hit.chunk.id] = entry
        }

        for (rank, hit) in semanticHits.enumerated() {
            let rrf = 1.0 / Double(60 + rank + 1)
            var entry = merged[hit.chunk.id] ?? (hit.chunk, 0, 0)
            entry.semanticScore += rrf
            merged[hit.chunk.id] = entry
        }

        return merged.values
            .sorted { lhs, rhs in
                let lhsTotal = lhs.keywordScore + lhs.semanticScore
                let rhsTotal = rhs.keywordScore + rhs.semanticScore
                if lhsTotal == rhsTotal {
                    return lhs.chunk.id < rhs.chunk.id
                }
                return lhsTotal > rhsTotal
            }
            .prefix(limit)
            .map { $0 }
    }

    func isReady(for corpus: AFICorpus) -> Bool {
        do {
            try openIfNeeded()
            guard indexedCorpusVersion == corpus.version else { return false }
            return try chunkCount() == corpus.chunks.count
        } catch {
            return false
        }
    }

    // MARK: - Private

    private func openIfNeeded() throws {
        if database != nil { return }

        try FileManager.default.createDirectory(
            at: databaseURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )

        var handle: OpaquePointer?
        guard sqlite3_open(databaseURL.path, &handle) == SQLITE_OK,
              let handle else {
            throw AFISearchIndexError.openFailed
        }

        database = handle
        try execute("PRAGMA journal_mode=WAL;")
        try execute("PRAGMA synchronous=NORMAL;")

        // On a fresh database the metadata table doesn't exist yet; treat that as
        // "no index built" rather than a failure.
        if let version = (try? metadataValue(for: "corpus_version")) ?? nil {
            indexedCorpusVersion = version
        }
    }

    private func rebuild(
        corpus: AFICorpus,
        progress: (@Sendable (Double) -> Void)? = nil
    ) throws {
        try execute("BEGIN IMMEDIATE TRANSACTION;")

        do {
            try execute("DROP TABLE IF EXISTS chunks_fts;")
            try execute("DROP TABLE IF EXISTS chunks;")
            try execute("DROP TABLE IF EXISTS metadata;")

            try execute(
                """
                CREATE TABLE chunks (
                    id TEXT PRIMARY KEY NOT NULL,
                    publication_id TEXT NOT NULL,
                    publication TEXT NOT NULL,
                    title TEXT NOT NULL,
                    section TEXT,
                    page INTEGER,
                    text TEXT NOT NULL,
                    embedding BLOB NOT NULL
                );
                """
            )

            try execute(
                """
                CREATE VIRTUAL TABLE chunks_fts USING fts5(
                    chunk_id UNINDEXED,
                    text,
                    publication,
                    title,
                    section,
                    tokenize = 'porter unicode61'
                );
                """
            )

            try execute(
                """
                CREATE TABLE metadata (
                    key TEXT PRIMARY KEY NOT NULL,
                    value TEXT NOT NULL
                );
                """
            )

            let insertChunkSQL = """
                INSERT INTO chunks (id, publication_id, publication, title, section, page, text, embedding)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?);
                """
            let insertFTSSQL = """
                INSERT INTO chunks_fts (chunk_id, text, publication, title, section)
                VALUES (?, ?, ?, ?, ?);
                """

            let insertChunk = try prepare(insertChunkSQL)
            defer { sqlite3_finalize(insertChunk) }

            let insertFTS = try prepare(insertFTSSQL)
            defer { sqlite3_finalize(insertFTS) }

            let total = Double(max(corpus.chunks.count, 1))

            for (index, chunk) in corpus.chunks.enumerated() {
                // Embeddings start empty and are backfilled asynchronously so keyword
                // search becomes available within seconds.
                let embeddingData = Data()

                try bindText(insertChunk, index: 1, chunk.id)
                try bindText(insertChunk, index: 2, chunk.publicationID)
                try bindText(insertChunk, index: 3, chunk.publication)
                try bindText(insertChunk, index: 4, chunk.title)
                try bindOptionalText(insertChunk, index: 5, chunk.section)
                if let page = chunk.page {
                    sqlite3_bind_int(insertChunk, 6, Int32(page))
                } else {
                    sqlite3_bind_null(insertChunk, 6)
                }
                try bindText(insertChunk, index: 7, chunk.text)
                try bindBlob(insertChunk, index: 8, embeddingData)
                try stepDone(insertChunk)
                sqlite3_reset(insertChunk)

                try bindText(insertFTS, index: 1, chunk.id)
                try bindText(insertFTS, index: 2, AFISearchIndex.ftsDocumentText(for: chunk))
                try bindText(insertFTS, index: 3, chunk.publication)
                try bindText(insertFTS, index: 4, chunk.title)
                try bindOptionalText(insertFTS, index: 5, chunk.section)
                try stepDone(insertFTS)
                sqlite3_reset(insertFTS)

                if index % 25 == 0 || index == corpus.chunks.count - 1 {
                    progress?(Double(index + 1) / total)
                }
            }

            let metadataStatement = try prepare("INSERT INTO metadata (key, value) VALUES (?, ?);")
            defer { sqlite3_finalize(metadataStatement) }
            try bindText(metadataStatement, index: 1, "corpus_version")
            try bindText(metadataStatement, index: 2, corpus.version)
            try stepDone(metadataStatement)

            try execute("COMMIT;")
        } catch {
            try? execute("ROLLBACK;")
            throw error
        }
    }

    private func keywordSearch(
        query: String,
        limit: Int
    ) throws -> [(chunk: AFICorpusChunk, score: Double)] {
        for ftsQuery in AFISearchQueryBuilder.ftsQueries(from: query) {
            let hits = try runKeywordSearch(ftsQuery: ftsQuery, limit: limit)
            if !hits.isEmpty {
                return hits
            }
        }
        return []
    }

    private func runKeywordSearch(
        ftsQuery: String,
        limit: Int
    ) throws -> [(chunk: AFICorpusChunk, score: Double)] {
        let sql = """
            SELECT
                c.id, c.publication_id, c.publication, c.title, c.section, c.page, c.text,
                bm25(chunks_fts) AS rank
            FROM chunks_fts
            JOIN chunks c ON c.id = chunks_fts.chunk_id
            WHERE chunks_fts MATCH ?
            ORDER BY rank
            LIMIT ?;
            """

        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }

        try bindText(statement, index: 1, ftsQuery)
        sqlite3_bind_int(statement, 2, Int32(limit))

        var results: [(AFICorpusChunk, Double)] = []

        while sqlite3_step(statement) == SQLITE_ROW {
            guard let chunk = try readChunk(from: statement) else { continue }
            let rank = sqlite3_column_double(statement, 7)
            let score = 1.0 / (1.0 + max(0.0, rank))
            results.append((chunk, score))
        }

        return results
    }

    private func semanticSearch(
        query: String,
        limit: Int
    ) throws -> [(chunk: AFICorpusChunk, score: Double)] {
        guard let queryVector = AFIEmbeddingService.embed(query) else { return [] }

        let sql = """
            SELECT id, publication_id, publication, title, section, page, text, embedding
            FROM chunks
            WHERE length(embedding) > 4;
            """

        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }

        var scored: [(AFICorpusChunk, Double)] = []

        while sqlite3_step(statement) == SQLITE_ROW {
            guard let chunk = try readChunk(from: statement),
                  let embeddingData = readBlob(statement, column: 7),
                  !embeddingData.isEmpty,
                  let vector = AFIEmbeddingService.deserialize(embeddingData) else {
                continue
            }

            let similarity = Double(AFIEmbeddingMath.cosineSimilarity(queryVector, vector))
            if similarity > 0.2 {
                scored.append((chunk, similarity))
            }
        }

        return scored
            .sorted {
                if $0.1 == $1.1 { return $0.0.id < $1.0.id }
                return $0.1 > $1.1
            }
            .prefix(limit)
            .map { $0 }
    }

    private static let embeddingUnavailableSentinel = Data([0])

    private func pendingEmbeddingBatch(limit: Int) throws -> [AFICorpusChunk] {
        let sql = """
            SELECT id, publication_id, publication, title, section, page, text
            FROM chunks
            WHERE length(embedding) = 0
            ORDER BY id
            LIMIT ?;
            """

        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }

        sqlite3_bind_int(statement, 1, Int32(limit))

        var chunks: [AFICorpusChunk] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            if let chunk = try readChunk(from: statement) {
                chunks.append(chunk)
            }
        }
        return chunks
    }

    private func chunkCount() throws -> Int {
        let statement = try prepare("SELECT COUNT(*) FROM chunks;")
        defer { sqlite3_finalize(statement) }

        guard sqlite3_step(statement) == SQLITE_ROW else { return 0 }
        return Int(sqlite3_column_int(statement, 0))
    }

    private func metadataValue(for key: String) throws -> String? {
        let statement = try prepare("SELECT value FROM metadata WHERE key = ? LIMIT 1;")
        defer { sqlite3_finalize(statement) }

        try bindText(statement, index: 1, key)
        guard sqlite3_step(statement) == SQLITE_ROW,
              let cString = sqlite3_column_text(statement, 0) else {
            return nil
        }
        return String(cString: cString)
    }

    private func readChunk(from statement: OpaquePointer?) throws -> AFICorpusChunk? {
        guard let statement,
              let id = readText(statement, column: 0),
              let publicationID = readText(statement, column: 1),
              let publication = readText(statement, column: 2),
              let title = readText(statement, column: 3) else {
            return nil
        }

        let section = readOptionalText(statement, column: 4)
        let page = sqlite3_column_type(statement, 5) == SQLITE_NULL
            ? nil
            : Int(sqlite3_column_int(statement, 5))
        let text = readText(statement, column: 6) ?? ""

        return AFICorpusChunk(
            id: id,
            publicationID: publicationID,
            publication: publication,
            title: title,
            section: section,
            page: page,
            text: text
        )
    }

    private func prepare(_ sql: String) throws -> OpaquePointer? {
        guard let database else { throw AFISearchIndexError.prepareFailed }

        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(database, sql, -1, &statement, nil) == SQLITE_OK else {
            logSQLiteError(context: "prepare: \(sql.prefix(80))")
            throw AFISearchIndexError.prepareFailed
        }
        return statement
    }

    private func execute(_ sql: String) throws {
        guard let database else { throw AFISearchIndexError.executionFailed }
        guard sqlite3_exec(database, sql, nil, nil, nil) == SQLITE_OK else {
            logSQLiteError(context: "execute: \(sql.prefix(80))")
            throw AFISearchIndexError.executionFailed
        }
    }

    private func stepDone(_ statement: OpaquePointer?) throws {
        let result = sqlite3_step(statement)
        guard result == SQLITE_DONE else {
            logSQLiteError(context: "step returned \(result)")
            throw AFISearchIndexError.executionFailed
        }
    }

    private func logSQLiteError(context: String) {
        let message = database.map { String(cString: sqlite3_errmsg($0)) } ?? "no database handle"
        print("AFISearchIndex: SQLite error (\(message)) — \(context)")
    }

    private func bindText(_ statement: OpaquePointer?, index: Int32, _ value: String) throws {
        guard sqlite3_bind_text(statement, index, value, -1, unsafeBitCast(-1, to: sqlite3_destructor_type.self)) == SQLITE_OK else {
            logSQLiteError(context: "bind text at index \(index)")
            throw AFISearchIndexError.executionFailed
        }
    }

    private func bindOptionalText(_ statement: OpaquePointer?, index: Int32, _ value: String?) throws {
        if let value {
            try bindText(statement, index: index, value)
        } else {
            sqlite3_bind_null(statement, index)
        }
    }

    private func bindBlob(_ statement: OpaquePointer?, index: Int32, _ data: Data) throws {
        let result = data.withUnsafeBytes { buffer in
            sqlite3_bind_blob(
                statement,
                index,
                buffer.baseAddress,
                Int32(data.count),
                unsafeBitCast(-1, to: sqlite3_destructor_type.self)
            )
        }
        guard result == SQLITE_OK else {
            throw AFISearchIndexError.executionFailed
        }
    }

    private func readText(_ statement: OpaquePointer?, column: Int32) -> String? {
        guard let statement,
              let cString = sqlite3_column_text(statement, column) else {
            return nil
        }
        return String(cString: cString)
    }

    private func readOptionalText(_ statement: OpaquePointer?, column: Int32) -> String? {
        guard let statement,
              sqlite3_column_type(statement, column) != SQLITE_NULL else {
            return nil
        }
        return readText(statement, column: column)
    }

    private func readBlob(_ statement: OpaquePointer?, column: Int32) -> Data? {
        guard let statement,
              let bytes = sqlite3_column_blob(statement, column) else {
            return nil
        }
        let length = Int(sqlite3_column_bytes(statement, column))
        return Data(bytes: bytes, count: length)
    }

    private static func ftsDocumentText(for chunk: AFICorpusChunk) -> String {
        [chunk.title, chunk.publication, chunk.section, chunk.text]
            .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: "\n")
    }
}

enum AFISearchQueryBuilder {
    nonisolated static func ftsQueries(from rawQuery: String) -> [String] {
        let tokens = tokenize(rawQuery)
        guard !tokens.isEmpty else { return [] }

        if tokens.count == 1 {
            return ["\"\(escape(tokens[0]))\"*"]
        }

        let phrase = tokens.map(escape).joined(separator: " ")
        let andClause = tokens.map { "\"\(escape($0))\"*" }.joined(separator: " AND ")
        let orClause = tokens.map { "\"\(escape($0))\"*" }.joined(separator: " OR ")

        return [
            "\"\(phrase)\"*",
            andClause,
            orClause
        ]
    }

    nonisolated private static func tokenize(_ rawQuery: String) -> [String] {
        rawQuery
            .lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count >= 2 }
    }

    nonisolated private static func escape(_ token: String) -> String {
        token.replacingOccurrences(of: "\"", with: "")
    }
}

File: android/app/src/main/java/com/ryanladuca/myafbase/data/afi/AfiSearchIndex.kt
package com.ryanladuca.myafbase.data.afi

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.util.Log
import com.ryanladuca.myafbase.data.repository.AfiChunk
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File

data class AfiSearchHit(
    val chunk: AfiChunk,
    val keywordScore: Double,
    val semanticScore: Double,
) {
    val matchKind: AfiMatchKind
        get() = when {
            keywordScore > 0 && semanticScore > 0 -> AfiMatchKind.BOTH
            semanticScore > 0 -> AfiMatchKind.SEMANTIC
            else -> AfiMatchKind.KEYWORD
        }

    val totalScore: Double get() = keywordScore + semanticScore
}

enum class AfiMatchKind { KEYWORD, SEMANTIC, BOTH }

class AfiSearchIndex(context: Context) {
    private val dbFile = File(context.filesDir, "afi_search/search_index.sqlite").apply {
        parentFile?.mkdirs()
    }
    private val mutex = Mutex()
    private var ftsMode: FtsMode = FtsMode.UNKNOWN

    suspend fun rebuildIfNeeded(corpusVersion: String, chunks: List<AfiChunk>, onProgress: ((Double) -> Unit)? = null) =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                openDatabase().use { db ->
                    if (!hasRequiredTables(db)) {
                        rebuild(db, corpusVersion, chunks, onProgress)
                        return@withLock
                    }
                    val existingVersion = metadataValue(db, "corpus_version")
                    val count = chunkCount(db)
                    if (existingVersion == corpusVersion && count == chunks.size) {
                        ftsMode = readFtsMode(db)
                        return@withLock
                    }
                    rebuild(db, corpusVersion, chunks, onProgress)
                }
            }
        }

    suspend fun pendingEmbeddingCount(): Int = withContext(Dispatchers.IO) {
        mutex.withLock {
            openDatabase().use { db ->
                if (!hasChunksTable(db)) return@withLock 0
                pendingEmbeddingCountLocked(db)
            }
        }
    }

    suspend fun backfillEmbeddings(
        batchSize: Int = 256,
        onProgress: ((Double) -> Unit)? = null,
    ) = withContext(Dispatchers.IO) {
        mutex.withLock {
            openDatabase().use { db ->
                if (!hasChunksTable(db)) return@withLock
                val totalPending = pendingEmbeddingCountLocked(db)
                if (totalPending == 0) {
                    onProgress?.invoke(1.0)
                    return@withLock
                }
                var completed = 0
                while (true) {
                    val batch = pendingEmbeddingBatch(db, batchSize)
                    if (batch.isEmpty()) break
                    val texts = batch.map { ftsDocumentText(it) }
                    val vectors = AfiEmbeddingService.embedAll(texts) { done, _ ->
                        val partial = completed + done
                        onProgress?.invoke((partial.toDouble() / totalPending).coerceAtMost(1.0))
                    }
                    db.beginTransaction()
                    try {
                        val statement = db.compileStatement("UPDATE chunks SET embedding = ? WHERE id = ?")
                        batch.forEachIndexed { index, chunk ->
                            val data = if (vectors[index].isEmpty()) EMBEDDING_UNAVAILABLE_SENTINEL else vectors[index]
                            statement.clearBindings()
                            statement.bindBlob(1, data)
                            statement.bindString(2, chunk.id)
                            statement.executeUpdateDelete()
                        }
                        db.setTransactionSuccessful()
                    } finally {
                        db.endTransaction()
                    }
                    completed += batch.size
                    onProgress?.invoke((completed.toDouble() / totalPending).coerceAtMost(1.0))
                }
            }
        }
    }

    suspend fun search(query: String, limit: Int = 12): List<AfiSearchHit> = withContext(Dispatchers.IO) {
        mutex.withLock {
            openDatabase().use { db ->
                if (!hasRequiredTables(db)) return@withLock emptyList()
                val trimmed = query.trim()
                if (trimmed.length < 2) return@withLock emptyList()

                val keywordHits = keywordSearch(db, trimmed, kotlin.math.max(limit * 4, 24))
                val semanticHits = if (hasEmbeddings(db)) {
                    semanticSearch(db, trimmed, kotlin.math.max(limit * 4, 24))
                } else {
                    emptyList()
                }

                val merged = linkedMapOf<String, AfiSearchHit>()
                keywordHits.forEachIndexed { rank, (chunk, _) ->
                    val rrf = 1.0 / (60 + rank + 1)
                    val existing = merged[chunk.id]
                    merged[chunk.id] = if (existing == null) {
                        AfiSearchHit(chunk, rrf, 0.0)
                    } else {
                        existing.copy(keywordScore = existing.keywordScore + rrf)
                    }
                }
                semanticHits.forEachIndexed { rank, (chunk, _) ->
                    val rrf = 1.0 / (60 + rank + 1)
                    val existing = merged[chunk.id]
                    merged[chunk.id] = if (existing == null) {
                        AfiSearchHit(chunk, 0.0, rrf)
                    } else {
                        existing.copy(semanticScore = existing.semanticScore + rrf)
                    }
                }
                merged.values.sortedWith(
                    compareByDescending<AfiSearchHit> { it.totalScore }
                        .thenBy { it.chunk.id },
                ).take(limit)
            }
        }
    }

    private fun openDatabase(): SQLiteDatabase =
        SQLiteDatabase.openDatabase(dbFile.absolutePath, null, SQLiteDatabase.CREATE_IF_NECESSARY)

    private fun rebuild(
        db: SQLiteDatabase,
        corpusVersion: String,
        chunks: List<AfiChunk>,
        onProgress: ((Double) -> Unit)?,
    ) {
        ftsMode = detectFtsMode(db)
        db.beginTransaction()
        try {
            db.execSQL("DROP TABLE IF EXISTS chunks_fts")
            db.execSQL("DROP TABLE IF EXISTS chunks")
            db.execSQL("DROP TABLE IF EXISTS metadata")
            db.execSQL(
                """
                CREATE TABLE chunks (
                    id TEXT PRIMARY KEY NOT NULL,
                    publication_id TEXT NOT NULL,
                    publication TEXT NOT NULL,
                    title TEXT NOT NULL,
                    section TEXT,
                    page INTEGER,
                    text TEXT NOT NULL,
                    embedding BLOB NOT NULL
                )
                """.trimIndent(),
            )
            createFtsTable(db, ftsMode)
            db.execSQL(
                """
                CREATE TABLE metadata (
                    key TEXT PRIMARY KEY NOT NULL,
                    value TEXT NOT NULL
                )
                """.trimIndent(),
            )

            val total = maxOf(chunks.size, 1)
            val chunkInsert = db.compileStatement(
                """
                INSERT INTO chunks (id, publication_id, publication, title, section, page, text, embedding)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """.trimIndent(),
            )
            val ftsInsert = db.compileStatement(
                """
                INSERT INTO chunks_fts (chunk_id, text, publication, title, section)
                VALUES (?, ?, ?, ?, ?)
                """.trimIndent(),
            )

            chunks.forEachIndexed { index, chunk ->
                chunkInsert.clearBindings()
                chunkInsert.bindString(1, chunk.id)
                chunkInsert.bindString(2, chunk.publicationId)
                chunkInsert.bindString(3, chunk.publication)
                chunkInsert.bindString(4, chunk.title)
                if (chunk.section != null) {
                    chunkInsert.bindString(5, chunk.section)
                } else {
                    chunkInsert.bindNull(5)
                }
                chunkInsert.bindLong(6, chunk.page.toLong())
                chunkInsert.bindString(7, chunk.text)
                chunkInsert.bindBlob(8, ByteArray(0))
                chunkInsert.executeInsert()

                ftsInsert.clearBindings()
                ftsInsert.bindString(1, chunk.id)
                ftsInsert.bindString(2, ftsDocumentText(chunk))
                ftsInsert.bindString(3, chunk.publication)
                ftsInsert.bindString(4, chunk.title)
                if (chunk.section != null) {
                    ftsInsert.bindString(5, chunk.section)
                } else {
                    ftsInsert.bindNull(5)
                }
                ftsInsert.executeInsert()

                if (index % 50 == 0 || index == chunks.lastIndex) {
                    onProgress?.invoke((index + 1).toDouble() / total)
                }
            }

            db.execSQL("INSERT INTO metadata (key, value) VALUES (?, ?)", arrayOf("corpus_version", corpusVersion))
            db.execSQL("INSERT INTO metadata (key, value) VALUES (?, ?)", arrayOf("fts_mode", ftsMode.name))
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    private fun detectFtsMode(db: SQLiteDatabase): FtsMode {
        if (ftsMode != FtsMode.UNKNOWN) return ftsMode
        return if (canCreateFts5(db)) FtsMode.FTS5 else FtsMode.FTS4
    }

    private fun canCreateFts5(db: SQLiteDatabase): Boolean = runCatching {
        db.execSQL("CREATE VIRTUAL TABLE IF NOT EXISTS fts5_probe USING fts5(text)")
        db.execSQL("DROP TABLE IF EXISTS fts5_probe")
        true
    }.getOrDefault(false)

    private fun createFtsTable(db: SQLiteDatabase, mode: FtsMode) {
        when (mode) {
            FtsMode.FTS5 -> db.execSQL(
                """
                CREATE VIRTUAL TABLE chunks_fts USING fts5(
                    chunk_id UNINDEXED,
                    text,
                    publication,
                    title,
                    section,
                    tokenize = 'porter unicode61'
                )
                """.trimIndent(),
            )
            FtsMode.FTS4 -> db.execSQL(
                """
                CREATE VIRTUAL TABLE chunks_fts USING fts4(
                    chunk_id,
                    text,
                    publication,
                    title,
                    section,
                    tokenize=porter
                )
                """.trimIndent(),
            )
            FtsMode.UNKNOWN -> error("FTS mode not resolved")
        }
    }

    private fun keywordSearch(db: SQLiteDatabase, query: String, limit: Int): List<Pair<AfiChunk, Double>> {
        for (ftsQuery in AfiSearchQueryBuilder.ftsQueries(query)) {
            val hits = runKeywordSearch(db, ftsQuery, limit)
            if (hits.isNotEmpty()) return hits
        }
        return emptyList()
    }

    private fun runKeywordSearch(db: SQLiteDatabase, ftsQuery: String, limit: Int): List<Pair<AfiChunk, Double>> {
        val sql = when (ftsMode) {
            FtsMode.FTS4 -> """
                SELECT c.id, c.publication_id, c.publication, c.title, c.section, c.page, c.text
                FROM chunks_fts
                JOIN chunks c ON c.id = chunks_fts.chunk_id
                WHERE chunks_fts MATCH ?
                LIMIT ?
            """.trimIndent()
            else -> """
                SELECT c.id, c.publication_id, c.publication, c.title, c.section, c.page, c.text,
                       bm25(chunks_fts) AS rank
                FROM chunks_fts
                JOIN chunks c ON c.id = chunks_fts.chunk_id
                WHERE chunks_fts MATCH ?
                ORDER BY rank
                LIMIT ?
            """.trimIndent()
        }
        val results = mutableListOf<Pair<AfiChunk, Double>>()
        return try {
            db.rawQuery(sql, arrayOf(ftsQuery, limit.toString())).use { cursor ->
                var rank = 0
                while (cursor.moveToNext()) {
                    val chunk = readChunk(cursor) ?: continue
                    val score = if (ftsMode == FtsMode.FTS4) {
                        rank += 1
                        1.0 / rank.toDouble()
                    } else {
                        val bm25Rank = cursor.getDouble(7)
                        1.0 / (1.0 + maxOf(0.0, bm25Rank))
                    }
                    results += chunk to score
                }
            }
            results
        } catch (error: SQLiteException) {
            Log.w(TAG, "FTS query failed: $ftsQuery", error)
            emptyList()
        }
    }

    private fun semanticSearch(db: SQLiteDatabase, query: String, limit: Int): List<Pair<AfiChunk, Double>> {
        val queryVector = AfiEmbeddingService.embed(query) ?: return emptyList()
        val sql = """
            SELECT id, publication_id, publication, title, section, page, text, embedding
            FROM chunks
            WHERE length(embedding) > 4
        """.trimIndent()
        val scored = mutableListOf<Pair<AfiChunk, Double>>()
        db.rawQuery(sql, null).use { cursor ->
            while (cursor.moveToNext()) {
                val chunk = readChunk(cursor) ?: continue
                val embedding = cursor.getBlob(7) ?: continue
                if (embedding.contentEquals(EMBEDDING_UNAVAILABLE_SENTINEL)) continue
                val vector = AfiEmbeddingService.deserialize(embedding) ?: continue
                val similarity = AfiEmbeddingMath.cosineSimilarity(queryVector, vector).toDouble()
                if (similarity > 0.2) scored += chunk to similarity
            }
        }
        return scored.sortedWith(compareByDescending<Pair<AfiChunk, Double>> { it.second }.thenBy { it.first.id })
            .take(limit)
    }

    private fun hasChunksTable(db: SQLiteDatabase): Boolean = runCatching {
        db.rawQuery(
            "SELECT name FROM sqlite_master WHERE type='table' AND name='chunks'",
            null,
        ).use { it.moveToFirst() }
    }.getOrDefault(false)

    private fun hasRequiredTables(db: SQLiteDatabase): Boolean = runCatching {
        db.rawQuery(
            "SELECT name FROM sqlite_master WHERE type='table' AND name IN ('chunks', 'chunks_fts', 'metadata')",
            null,
        ).use { cursor ->
            var count = 0
            while (cursor.moveToNext()) count++
            count == 3
        }
    }.getOrDefault(false)

    private fun hasEmbeddings(db: SQLiteDatabase): Boolean = runCatching {
        db.rawQuery("SELECT 1 FROM chunks WHERE length(embedding) > 4 LIMIT 1", null).use { it.moveToFirst() }
    }.getOrDefault(false)

    private fun pendingEmbeddingCountLocked(db: SQLiteDatabase): Int =
        db.rawQuery("SELECT COUNT(*) FROM chunks WHERE length(embedding) = 0", null).use { cursor ->
            if (cursor.moveToFirst()) cursor.getInt(0) else 0
        }

    private fun pendingEmbeddingBatch(db: SQLiteDatabase, limit: Int): List<AfiChunk> {
        val sql = """
            SELECT id, publication_id, publication, title, section, page, text
            FROM chunks
            WHERE length(embedding) = 0
            LIMIT ?
        """.trimIndent()
        val batch = mutableListOf<AfiChunk>()
        db.rawQuery(sql, arrayOf(limit.toString())).use { cursor ->
            while (cursor.moveToNext()) {
                readChunk(cursor)?.let { batch += it }
            }
        }
        return batch
    }

    private fun chunkCount(db: SQLiteDatabase): Int =
        db.rawQuery("SELECT COUNT(*) FROM chunks", null).use { cursor ->
            if (cursor.moveToFirst()) cursor.getInt(0) else 0
        }

    private fun metadataValue(db: SQLiteDatabase, key: String): String? =
        db.rawQuery("SELECT value FROM metadata WHERE key = ?", arrayOf(key)).use { cursor ->
            if (cursor.moveToFirst()) cursor.getString(0) else null
        }

    private fun readFtsMode(db: SQLiteDatabase): FtsMode =
        metadataValue(db, "fts_mode")?.let { runCatching { FtsMode.valueOf(it) }.getOrNull() }
            ?: FtsMode.FTS5

    private fun readChunk(cursor: android.database.Cursor): AfiChunk? = runCatching {
        AfiChunk(
            id = cursor.getString(0),
            publicationId = cursor.getString(1),
            publication = cursor.getString(2),
            title = cursor.getString(3),
            section = cursor.getString(4),
            page = if (cursor.isNull(5)) 0 else cursor.getInt(5),
            text = cursor.getString(6),
        )
    }.getOrNull()

    private fun ftsDocumentText(chunk: AfiChunk): String =
        listOfNotNull(chunk.publication, chunk.title, chunk.section, chunk.text)
            .joinToString("\n")

    private enum class FtsMode { UNKNOWN, FTS5, FTS4 }

    companion object {
        private const val TAG = "AfiSearchIndex"
        private val EMBEDDING_UNAVAILABLE_SENTINEL = byteArrayOf(0)
    }
}
